src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -1226,3 +1264,39 @@ std::string SotaUptaneClient::secondaryTreehubCredentials() const {
 Uptane::LazyTargetsList SotaUptaneClient::allTargets() const {
   return Uptane::LazyTargetsList(images_repo, storage, uptane_fetcher);
 }
+
+void SotaUptaneClient::checkAndUpdatePendingSecondaries() {
+  // TODO: think of another alternatives to inform Primary about installation result on Secondary ECUs (reboot case)
+  std::vector<std::pair<Uptane::EcuSerial, Uptane::Hash>> pending_ecus;
+  storage->getPendingEcus(&pending_ecus);
+
+  for (const auto &pending_ecu : pending_ecus) {
+    if (primaryEcuSerial() == pending_ecu.first) {
+      continue;
+    }
+    auto &sec = secondaries[pending_ecu.first];
+    const auto &manifest = sec->getManifest();
+    if (!manifest.verifySignature(sec->getPublicKey())) {
+      LOG_ERROR << "Invalid signrature of the manifest reported by secondary: "
+                << " serial: " << pending_ecu.first << " manifest: " << manifest;
+      // what should we do in this case ?
+      continue;
+    }
+    auto current_ecu_hash = manifest.installedImageHash();
+    if (pending_ecu.second == current_ecu_hash) {
+      LOG_INFO << "The pending update " << current_ecu_hash << " has been installed on " << pending_ecu.first;
+      boost::optional<Uptane::Target> pending_version;
+      if (storage->loadInstalledVersions(pending_ecu.first.ToString(), nullptr, &pending_version)) {
+        storage->saveEcuInstallationResult(pending_ecu.first,
+                                           data::InstallationResult(data::ResultCode::Numeric::kOk, ""));
+
+        storage->saveInstalledVersion(pending_ecu.first.ToString(), *pending_version,
+                                      InstalledVersionUpdateMode::kCurrent);
+
+        report_queue->enqueue(std_::make_unique<EcuInstallationCompletedReport>(
+            pending_ecu.first, director_repo.getCorrelationId(), true));
+        computeDeviceInstallationResult(nullptr, director_repo.getCorrelationId());
```

Comment:
It looks to me like some of this code is quite similar to `AssembleManifest()`. Can they reuse more code?

Reply:
It doesn't look similar to me, the code update the currently installed version for secondary ECUs while AssembleManifest() just assembles a device manifest.

Reply:
Hmmm, now I don't know what I was thinking of there. Never mind.